### PR TITLE
Add Spark SQL support

### PR DESCRIPTION
**Pull request by gengliangwang**
# Add Spark SQL support 
* Add Spark SQL support. It can connect to Spark via building a local/remote SparkSession.
* Include a notebook example

I tried some complicated queries (window function, table joins), and the tool works well. 
Compared to the [Spark Dataframe agent](https://python.langchain.com/en/latest/modules/agents/toolkits/examples/spark.html), this tool is able to generate queries across multiple tables.

### CODE REVIEW

**Comment by gengliangwang** [comment]
Note: There was an approach based on `SQLDatabase`. But @dev2049 suggests not inheriting from SQLDatabase.
https://github.com/hwchase17/langchain/pull/4381

File: langchain/tools/spark_sql/tool.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    @root_validator(pre=True)
+    def initialize_llm_chain(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        if "llm_chain" not in values:
+            values["llm_chain"] = LLMChain(
+                llm=values.get("llm"),
+                prompt=PromptTemplate(
+                    template=QUERY_CHECKER, input_variables=["query"]
+                ),
+            )
+
```
Comment (skcoirz) [inline langchain/tools/spark_sql/tool.py:130]: if we force users to use the default prompt, I think it makes sense here and then I would change the error message below to be more specific.

**Comment by gengliangwang** [comment]
@skcoirz Thanks for help updating this one!

**Comment by skcoirz** [comment]
> @skcoirz Thanks for help updating this one!

yeah, sure thing! I tested this. The new query checker is really powerful! It solved the previous concern of AnalysisException. Thank you so much for adding this! During the test, I noticed a few more opportunities. I have added them to our spreadsheet. Happy to chat more when you have time! Have a good weekend! :D

File: langchain/agents/agent_toolkits/spark_sql/base.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+from langchain.agents.mrkl.base import ZeroShotAgent
+from langchain.base_language import BaseLanguageModel
+from langchain.callbacks.base import BaseCallbackManager
+from langchain.chains.llm import LLMChain
+from langchain.chat_models import ChatOpenAI
+from langchain.memory import ConversationBufferMemory
+from langchain.spark_sql import SparkSQL
+
+
+def create_spark_analytics_agent_verified(
```
Comment (gengliangwang) [inline langchain/agents/agent_toolkits/spark_sql/base.py:27]: @skcoirz Having two methods for creating agents looks confusing.

File: langchain/agents/agent_toolkits/spark_sql/base.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    schema: str,
+    allow_freestyle: bool = False,
+) -> AgentExecutor:
+    """
+    Construct a spark analytics agent with ChatGPT-4,
+    which is verified during development.
+    """
+    spark_sql = SparkSQL(schema=schema)
+    llm = ChatOpenAI(temperature=0, model_name="gpt-4")
```
Comment (gengliangwang) [inline langchain/agents/agent_toolkits/spark_sql/base.py:36]: Let's not bind gpt-4 here. Langchain is supposed to be general.

Comment (gengliangwang) [reply langchain/agents/agent_toolkits/spark_sql/base.py]: @skcoirz I am reverting the related changes. You can still find the code in the git history.

Comment (skcoirz) [reply langchain/agents/agent_toolkits/spark_sql/base.py]: sure, I’m good with it. There is a tradeoff. We can chat more later.

**Comment by skcoirz** [comment]
Moved the rest new features to a new PR on top of branch. (https://github.com/hwchase17/langchain/pull/4672)

**Comment by gengliangwang** [comment]
cc @vowelparrot @hwchase17 could you review this one? The new agent is helpful for the Apache Spark community.

File: langchain/agents/mrkl/output_parser.py [hunk 4]
```diff
@@ -22,6 +22,9 @@ def parse(self, text: str) -> Union[AgentAction, AgentFinish]:
             r"Action\s*\d*\s*:[\s]*(.*?)[\s]*Action\s*\d*\s*Input\s*\d*\s*:[\s]*(.*)"
         )
         match = re.search(regex, text, re.DOTALL)
+        if not match:  # try again with another format as LLM output is not stable
```
Comment (hwchase17) [inline langchain/agents/mrkl/output_parser.py:25]: this seems a bit separate? could we handle/discuss in a separate pr?

**Review by hwchase17 — COMMENTED** [review]
one comment around some part looking a bit out of place, but otherwise looks good to me! if you dont get around to it, we can also remove the commented bit and merge in tonight @dev2049

File: langchain/agents/mrkl/output_parser.py (hunk 4 above)
Comment (gengliangwang) [reply langchain/agents/mrkl/output_parser.py]: @hwchase17 Sure. I just removed it.
cc @skcoirz Please create another PR for the parser changes and explain why we need it.

**Comment by gengliangwang** [comment]
I just did a final check before merging. 
There is a bug in the memory support. I reverted it to make this first version simple and robust. Discuss with @skcoirz offline and he will create another PR for general support for Agents. 
I also verified by rerunning the notebook. It works great.

File: langchain/spark_sql.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
+from __future__ import annotations
```
Comment (hwchase17) [inline langchain/spark_sql.py:1]: nit: lets move this to utilities

Comment (dev2049) [reply langchain/spark_sql.py:1]: (as in the whole file should go to langchain/utilities/spark_sql, not that specific line)

Comment (hwchase17) [reply langchain/spark_sql.py:1]: yeah that whole file

Comment (gengliangwang) [reply langchain/spark_sql.py:1]: Make sense. This is done.

**Comment by gengliangwang** [comment]
@hwchase17 @dev2049 @skcoirz Thanks for reviewing this!

**Pull request merged**